libraries/chain/transaction_context.cpp

```diff
@@ -137,72 +137,84 @@ namespace eosio { namespace chain {
+            eager_net_limit = new_eager_net_limit;
+            net_limit_due_to_block = false;
+         }
 
-      // Possibly limit deadline if the duration accounts can be billed for (+ a subjective leeway) does not exceed current delta
-      if( (fc::microseconds(account_cpu_limit) + leeway) <= (_deadline - start) ) {
-         _deadline = start + fc::microseconds(account_cpu_limit) + leeway;
-         billing_timer_exception_code = leeway_deadline_exception::code_value;
-      }
+         // Possibly limit deadline if the duration accounts can be billed for (+ a subjective leeway) does not exceed current delta
+         if( (fc::microseconds(account_cpu_limit) + leeway) <= (_deadline - start) ) {
+            _deadline = start + fc::microseconds(account_cpu_limit) + leeway;
+            billing_timer_exception_code = leeway_deadline_exception::code_value;
+         }
 
-      // Possibly limit deadline to subjective max_transaction_time
-      if( max_transaction_time_subjective != fc::microseconds::maximum() && (start + max_transaction_time_subjective) <= _deadline ) {
-         _deadline = start + max_transaction_time_subjective;
-         billing_timer_exception_code = tx_cpu_usage_exceeded::code_value;
-      }
+         // Possibly limit deadline to subjective max_transaction_time
+         if( max_transaction_time_subjective != fc::microseconds::maximum() && (start + max_transaction_time_subjective) <= _deadline ) {
+            _deadline = start + max_transaction_time_subjective;
+            billing_timer_exception_code = tx_cpu_usage_exceeded::code_value;
+         }
 
-      // Possibly limit deadline to caller provided wall clock block deadline
-      if( block_deadline < _deadline ) {
-         _deadline = block_deadline;
-         billing_timer_exception_code = deadline_exception::code_value;
-      }
+         // Possibly limit deadline to caller provided wall clock block deadline
+         if( block_deadline < _deadline ) {
+            _deadline = block_deadline;
+            billing_timer_exception_code = deadline_exception::code_value;
+         }
 
-      if( !explicit_billed_cpu_time ) {
-         int64_t validate_account_cpu_limit = account_cpu_limit - subjective_cpu_bill_us + leeway.count(); // Add leeway to allow powerup
-         // Possibly limit deadline to account subjective cpu left
-         if( subjective_cpu_bill_us > 0 && (start + fc::microseconds(validate_account_cpu_limit) < _deadline) ) {
-            _deadline = start + fc::microseconds(validate_account_cpu_limit);
-            billing_timer_exception_code = tx_cpu_usage_exceeded::code_value;
+         // Explicit billed_cpu_time_us should be used, block_deadline will be maximum unless in test code
+         if( explicit_billed_cpu_time ) {
+            _deadline = block_deadline;
+            deadline_exception_code = deadline_exception::code_value;
+         } else {
+            deadline_exception_code = billing_timer_exception_code;
          }
```

Comment:
Thanks @heifner for your careful review! Please ignore the above emoji. It was clicked by mistake.

Reply:
Yes, we will add a new node operator limit for read-only transactions when multi-threaded is implemented shortly. That is why I leave on-chain `max_transaction_cpu_usage` and block cpu limit for the worst case protection for now.

On another note, do you think I should go back to the original separate `init_for_read_only_trx`, not polluting the main `init`?

Reply:
Should this PR be targeted to a feature branch then? Not sure about merging this into `main` until it is complete. Maybe we should put a hard-coded limit for now. 150ms is a rather large amount of time.

No, I still think having this all in `init` is easier to follow.

Reply:
Read-only transactions can be a useful separate feature. I will add a hard-coded limit. It would be good to get into `main` to soak it.

I will keep all initializations in `init`.

Reply:
The code is still duplicated.

Reply:
I am so sorry for this merge error.